// riot/esm/api/__.js
/* Riot v9.4.9, @license MIT */
import bindingTypes from '../dependencies/@riotjs/util/binding-types.js';
import { PARENT_KEY_SYMBOL, DOM_COMPONENT_INSTANCE_PROPERTY, PLUGINS_SET, COMPONENTS_IMPLEMENTATION_MAP, ATTRIBUTES_KEY_SYMBOL, IS_COMPONENT_UPDATING, STATE_KEY, PROPS_KEY } from '../dependencies/@riotjs/util/constants.js';
import expressionTypes from '../dependencies/@riotjs/util/expression-types.js';
import { createExpression as create$4, createBinding as create$1, template as create } from '../dependencies/@riotjs/dom-bindings/dist/dom-bindings.js';
import cssManager from '../core/css-manager.js';

// expose some internal stuff that might be used from external tools
const __ = {
  cssManager,
  DOMBindings: {
    template: create,
    createBinding: create$1,
    createExpression: create$4,
    bindingTypes,
    expressionTypes,
  },
  globals: {
    PROPS_KEY,
    STATE_KEY,
    IS_COMPONENT_UPDATING,
    ATTRIBUTES_KEY_SYMBOL,
    COMPONENTS_IMPLEMENTATION_MAP,
    PLUGINS_SET,
    DOM_COMPONENT_INSTANCE_PROPERTY,
    PARENT_KEY_SYMBOL,
  },
};

export { __ };
